ui/jobs/components/base.ts

```diff
@@ -0,0 +1,114 @@
+import ResourceBar from '../../../resources/resourcebar';
+import TimerBar from '../../../resources/timerbar';
+import TimerBox from '../../../resources/timerbox';
+import { JobDetail } from '../../../types/event';
+import { Job } from '../../../types/job';
+import { NetMatches } from '../../../types/net_matches';
+import { Bars } from '../jobs';
+import Player, { Stats } from '../player';
+
+export class BaseComponent<T extends Job> {
+  bars: Bars;
+  player: Player;
+
+  constructor(bars: Bars) {
+    this.bars = bars;
+    this.player = bars.player;
+  }
+
+  addCustomBar(element: HTMLElement): HTMLElement {
+    const container = this.bars.addJobBarContainer();
+    container.appendChild(element);
+    return container;
+  }
+
+  addProcBox(options: {
+    id: string;
+    fgColor?: string;
+    threshold?: number;
+    scale?: number;
+    notifyWhenExpired?: boolean;
+  }): TimerBox {
+    return this.bars.addProcBox(options) as TimerBox;
+  }
+
+  addResourceBar(options: {
+    id: string;
+    fgColor?: string;
+    maxvalue?: string;
+  }): ResourceBar {
+    return this.bars.addResourceBar(options) as ResourceBar;
+  }
+
+  addResourceBox(options: {
+    classList?: string[];
+  }): HTMLDivElement {
+    return this.bars.addResourceBox(options);
+  }
+
+  addTimerBar(options: {
+    id: string;
+    fgColor?: string;
+  }): TimerBar {
+    return this.bars.addTimerBar(options);
+  }
+
+  // In order to provide full type definitions for inherited components work,
+  // we need to disable these rules.
```

Comment:
Is this something that `abstract` would fix?

Reply:
Non unfortunately.
I tried before but the `abstract` modifier requires inherited classes to implement all abstract methods, that usually a job component need several of them only instead.

Another approach that don't need to disable rules is to have a `ComponentInterface` that provides types definitions for all the methods but marks them as optional, and then all classes would `implement` it.

Reply:
Yeah, because these are all concrete classes in the downstream .ts files, something like this should work:

```ts
interface ComboComponent {
  /**
   * (override) called on combo state changed
   */
  onCombo(skill: string): void;
}

// ...

if (someComponent instanceof ComboComponent)
  someComponent.onCombo(...);
```

Alternatively, you could leave both rules enabled and define the functions as:

```ts
  /**
   * (override) called on combo state changed
   */
  onCombo(_skill: string): void {
    // This space intentionally left blank
  }
```

Reply:
If so, I have thought. How about using [Mixins](https://www.typescriptlang.org/docs/handbook/mixins.html) and inherit methods that are needed only?

Reply:
I don't think mixins really make sense here since we don't have any default behavior or extra properties that would benefit from being inherited here, everything needs to be redefined downstream anyways.

Mixins are more of a way to handle multiple class inheritance, rather than what's going on here.

Reply:
Yeah, that's why I mentioned it.
For example, the method `onCombo` that you mentioned has basically the same code across components, mixins can be a good way to replenish the lacking of "multiple inherits" in this case. What do you think?

Additionally, I am thinking to move code that related to specific jobs out of `jobs.js`. Such as `mpTick` for BLM, or `cpBar` for crafters, and so on. Do you know how to deal with these various mechanics perfectly?